src/Compilers/Core/Portable/Compilation.EmitStream.cs

```diff
@@ -0,0 +1,176 @@
+            {
+                Debug.Assert(emitStreamProvider != null);
+                Debug.Assert(strongNameProvider != null || emitStreamSignKind == EmitStreamSignKind.None);
+                _emitStreamProvider = emitStreamProvider;
+                _emitStreamSignKind = emitStreamSignKind;
+                _strongNameProvider = strongNameProvider;
+            }
+
+            internal Func<Stream> GetCreateStreamFunc(DiagnosticBag diagnostics)
+            {
+                return () => CreateStream(diagnostics);
+            }
+
+            internal void Dispose()
+            {
+                _tempInfo?.tempStream.Dispose();
+                _tempInfo = null;
+
+                // The _stream value is deliberately excluded from being disposed here. That value is not 
+                // owned by this type.
+                _stream = null;
+            }
+
+            /// <summary>
+            /// Create the stream which should be used for Emit. This should only be called one time.
+            /// </summary>
+            private Stream CreateStream(DiagnosticBag diagnostics)
+            {
+                Debug.Assert(_stream == null);
+                Debug.Assert(diagnostics != null);
+
+                if (diagnostics.HasAnyErrors())
+                {
+                    return null;
+                }
+
+                _stream = _emitStreamProvider.GetOrCreateStream(diagnostics);
+                if (_stream == null)
+                {
+                    Debug.Assert(diagnostics.HasAnyErrors());
+                    return null;
+                }
+
+                // If the current strong name provider is the Desktop version, signing can only be done to on-disk files.
+                // If this binary is configured to be signed, create a temp file, output to that
+                // then stream that to the stream that this method was called with. Otherwise output to the
+                // stream that this method was called with.
+                if (_emitStreamSignKind == EmitStreamSignKind.SignedWithFile)
+                {
+                    Debug.Assert(_strongNameProvider != null);
+
+                    try
+                    {
+                        var fileSystem = _strongNameProvider.FileSystem;
+                        Func<string, Stream> streamConstructor = path => fileSystem.CreateFileStream(path, FileMode.Create, FileAccess.ReadWrite, FileShare.ReadWrite);
+
+                        var tempDir = fileSystem.GetTempPath();
+                        var tempFilePath = Path.Combine(tempDir, Guid.NewGuid().ToString("N"));
+                        var tempStream = FileUtilities.CreateFileStreamChecked(streamConstructor, tempFilePath);
+
+                        _tempInfo = (tempStream, tempFilePath);
+                        return tempStream;
+                    }
+                    catch (IOException e)
+                    {
+                        throw new Cci.PeWritingException(e);
```

Comment:
Consider recording that there was an error in some manner, so `Complete` can assert that, for `EmitStreamSignKind.SignedWithFile`, we either have temp info or there was an error.

Reply:
K. Think I can just re-arrange this a bit and assert on the exisiting state. 

---
In reply to: [245374861](https://github.com/dotnet/roslyn/pull/31986#discussion_r245374861) [](ancestors = 245374861)